### PR TITLE
Reconcile test cases w/ latest spec changes and fixes from latest connect-go

### LINKED ISSUE #810: HTTP error codes should be considered when error response is malformed in unary requests
Trace:
```
FAILED: Connect Error and End-Stream/HTTPVersion:1/Compression:COMPRESSION_BR/TLS:false/error/null:
        actual error code 16 (unauthenticated) does not match expected code 2 (unknown)
---- HTTP Trace ----
 request>     0.000ms POST http://.../connectrpc.conformance.v1.ConformanceService/Unary HTTP/1.1
 request>             Accept-Encoding: br
 request>             Connect-Protocol-Version: 1
 request>             Connection: keep-alive
 request>             Content-Encoding: br
 request>             Content-Type: application/proto
 request>             User-Agent: connect-es/1.3.0
 request>             X-Expect-Codec: 1
 request>             X-Expect-Compression: 3
 request>             X-Expect-Http-Method: POST
 request>             X-Expect-Http-Version: 1
 request>             X-Expect-Protocol: 1
 request>             X-Expect-Tls: false
 request>             X-Test-Case-Name: Connect Error and End-Stream/HTTPVersion:1/Compression:COMPRESSION_BR/TLS:false/error/null
 request>
 request>     0.408ms message #1: data: 53 bytes
 request>     0.409ms body end
response<     0.550ms 401 Unauthorized
response<             Content-Type: application/json
response<
response<     0.553ms message #1: data: 4 bytes
response<     0.554ms body end
--------------------
```

Here, the reference server returns a 401 HTTP status code but returns `null` body. This expects the error code thrown by the client to be `Unknown`. This should in fact be `Unauthenticated` because typically such errors can be returned by any proxy that handles auth, rate limiting etc...

The behavior should be to fallback to the HTTP status code if the error code is malformed in the response.

### LINKED ISSUE #813: Timeout test may cancel before server sends DEADLINE_EXCEEDED
The test cases:

- Timeouts/HTTPVersion:2/Protocol:PROTOCOL_GRPC/Codec:CODEC_PROTO/Compression:COMPRESSION_GZIP/TLS:false/bidi-stream/full-duplex
- Timeouts/HTTPVersion:2/Protocol:PROTOCOL_GRPC/Codec:CODEC_PROTO/Compression:COMPRESSION_IDENTITY/TLS:false/bidi-stream/full-duplex

are currently failing against my [gRPC-Java server implementation](https://github.com/bufbuild/conformance-grpc-java)) as follows:

```
FAILED: Timeouts/HTTPVersion:2/Protocol:PROTOCOL_GRPC/Codec:CODEC_PROTO/Compression:COMPRESSION_IDENTITY/TLS:false/bidi-stream/full-duplex:
actual error code 1 (canceled) does not match expected code 4 (deadline_exceeded)
---- HTTP Trace ----
request>     0.000ms POST http://:52863/connectrpc.conformance.v1.ConformanceService/BidiStream HTTP/1.1
request>             Accept-Encoding: identity
request>             Content-Type: application/grpc
request>             Grpc-Timeout: 199990u
request>             Te: trailers
request>             User-Agent: grpc-go-connect/1.15.0 (go1.22.0) connectconformance-referenceclient/v1.0.0-rc3
request>             X-Test-Case-Name: Timeouts/HTTPVersion:2/Protocol:PROTOCOL_GRPC/Codec:CODEC_PROTO/Compression:COMPRESSION_IDENTITY/TLS:false/bidi-stream/full-duplex
request>
request>     0.401ms message #1: prefix: flags=0, len=37
request>             message #1: data: 37/37 bytes
response<     4.125ms 200 OK
response<             Content-Type: application/grpc
response<             Grpc-Accept-Encoding: gzip
response<             Grpc-Encoding: identity
response<
request>   199.301ms canceled
```

It appears that the client cancels at 200ms while the server is still sleeping for 1500ms.

### PR DESCRIPTION
**Pull request by jhump**
This updates to the latest connect-go release and reconciles the test cases with the latest spec changes (from https://github.com/connectrpc/connectrpc.com/pull/130). It also enables some other tests that were waiting on connect-go fixes and cleans up the "known failing" lists based on these fixes and the new test cases. Finally, there are some tweaks to gRPC test cases, related to relaxing error code assertions.

This should be the last thing we need before releasing v1.0.0-rc4.

Resolves #810.

This also brings in [a fix in connect-go](https://github.com/connectrpc/connect-go/pull/709) that resolves #813.

### CODE REVIEW

File: testing/referenceclient-known-failing.txt [hunk 1]
```diff
@@ -1,11 +1 @@
...
-# both clients and servers, should be an "unimplemented" error.
-#
-# Search for "cardinality violation" in this doc:
-#   https://grpc.github.io/grpc/core/md_doc_statuscodes.html
-**/unary/multiple-responses
-**/unary/ok-but-no-response
-**/client-stream/multiple-responses
-**/client-stream/ok-but-no-response
\ No newline at end of file
+# W00T! There are no known failures in the reference client.
```
Comment (akshayjshah) [inline testing/referenceclient-known-failing.txt:1]: I'm on board with the w00t vibes, but our OSS vibe is typically much more buttoned-up 🤷🏽‍♂️ 

```suggestion
# There are no known failures in the reference client.
```

File: testing/referenceserver-known-failing.txt [hunk 2]
```diff
@@ -1,8 +1 @@
-# The entries for "cardinality violation" in the following doc indicate that
-# these cases should fail with "unimplemented":
-#   https://grpc.github.io/grpc/core/md_doc_statuscodes.html
-# But the grpc-go client instead fails with "unknown".
-**/unary/multiple-requests
-**/unary/no-request
-**/server-stream/multiple-requests
-**/server-stream/no-request
+# W00T! There are no known failures in the reference server.
```
Comment (akshayjshah) [inline testing/referenceserver-known-failing.txt:1]:
```suggestion
# There are no known failures in the reference server.
```

File: internal/app/connectconformance/testsuites/data/server_unexpected.yaml [hunk 3]
```diff
@@ -21,7 +18,14 @@ testCases:
...
+      - CODE_INTERNAL
+      - CODE_UNIMPLEMENTED
     expectedResponse:
-      httpStatusCode: 415
+      # Ideally, the server would return a 415. But it's not actually specified :/
+      # httpStatusCode: 415
       error:
-        code: CODE_UNKNOWN
\ No newline at end of file
+        code: CODE_UNKNOWN
```
Comment (akshayjshah) [inline internal/app/connectconformance/testsuites/data/server_unexpected.yaml:31]: IIRC the Connect protocol is very particular about this scenario - should we separate the gRPC and Connect test expectations?

**Review by akshayjshah — APPROVED** [review]
Some small nits and questions, but this looks good to me! If subsequent changes are limited to comments and potentially splitting the Connect & gRPC assertions for unexpected content types, fine to merge without another review.

File: internal/app/connectconformance/testsuites/data/server_unexpected.yaml (hunk 3 above)
Comment (jhump) [reply internal/app/connectconformance/testsuites/data/server_unexpected.yaml:31]: Sure. I was actually thinking about that this morning, too.

**Pull request merged**